### DISCUSSION TITLE
Pydantic V2 with fastapi

### DISCUSSION THREAD

**Discussion by pablocimatoteramot** [fastapi/fastapi#11770] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# the endpoint


@request_router.get("/{id}", response_model=GetOneRequestResponseSchema)
def get_one_by_id(id: str):
    """
    Get the status of an ETL job by its ID.

    Args:
        id (str): The ID of the ETL job.

    Returns:
        dict: A dictionary containing the job result.
    """
    logger.info(f"Fetching status for ETL job with ID: {id}")
    request = RequestService.get_one(id)
    return request


class BaseSchema(BaseModel):
    """Base Schema."""

    model_config = ConfigDict(from_attributes=True)

    @field_validator("*")
    def datetime_format(cls, value):
        if isinstance(value, datetime):
            value = value.replace(tzinfo=timezone.utc)
        return value


class BaseRequestSchema(BaseSchema):
    """Base Schema for Request."""
    pass


class BaseResponseSchema(BaseSchema):
    """Base Schema for Response."""

    status: str = Field(default="success")
    data: Any

    model_config = ConfigDict(from_attributes=True)

    @model_validator(mode="before")
    def format(cls, values):
        return {
            "status": "success",
            "data": values,
        }


class UniqueMainValuesSchemaOut(BaseSchema):
    value: str


class UniqueValueSchemaOut(BaseSchema):
    unique_values_count: Optional[int] = None
    dtype: Optional[str] = None
    has_more_than_10_values: Optional[bool] = True
    main_unique_values: Optional[List[UniqueMainValuesSchemaOut]] = None

# part of the schema out

class FindingsSchemaOut(UniqueValueSchemaOut):
    finding_type_id: str
    details: Dict = None

    @model_validator(mode="before")
    @classmethod
    def format_finding_output(cls, values):
        print(f"values: {values.__dict__}")
        values_dict = values.__dict__
        if isinstance(values.finding_type_id, str):
            values_dict["finding_type_id"] = FindingTypes(values_dict["finding_type_id"]).name
            if values_dict["finding_type_id"] == FindingTypes.FINDING_UNIQUE_VALUES.name:
                values_dict["details"] = {
                    "unique_values_count": values_dict.pop("unique_values_count", 0),
                    "dtype": values_dict.pop("dtype", ""),
                    "has_more_than_10_values": values_dict.pop("has_more_than_10_values", False),
                    "main_unique_values": [value.value for value in values_dict.pop("main_unique_values", [])]
                }
        return values_dict
```

### Description

I am trying to format a response model output with pydantic version 2, and i am not able to do it.
I used to do it with root_validator in the v1,

Something like:

```
        @root_validator
    def format_finding_output(cls, values):
        if isinstance(values.finding_type_id, str):
            values_dict["finding_type_id"] = FindingTypes(values["finding_type_id"]).name
            if values["finding_type_id"] == FindingTypes.FINDING_UNIQUE_VALUES.name:
                values["details"] = {
                    "unique_values_count": values.pop("unique_values_count", 0),
                    "dtype": values.pop("dtype", ""),
                    "has_more_than_10_values": values.pop("has_more_than_10_values", False),
                    "main_unique_values": [value.value for value in values.pop("main_unique_values", [])]
                }
        return values
```

        
        but in this version and with model_validator i am not able to replace the original atributes so i have an output like:
        
        `"findings": [
                  {
                    "unique_values_count": null,
                    "dtype": null,
                    "has_more_than_10_values": true,
                    "main_unique_values": null,
                    "finding_type_id": "FINDING_UNIQUE_VALUES",
                    "details": {
                      "unique_values_count": 10,
                      "dtype": "string",
                      "has_more_than_10_values": false,
                      "main_unique_values": [
                        "2402.0",
                        "TREFILACIÓN ROLLO",
                        "0",
                        "nan",
                        "TREFILACION ROLLOS",
                        "DEVOLUCION",
                        "Generación Manual",
                        "NOTA INGRESO",
                        "ENDEREZADO",
                        "nan"
                      ]
                    }
                  },`
                  
                  
                  
        
        

### Operating System

Linux

### Operating System Details

fastapi==0.111.0
pydantic==2.7.1

### FastAPI Version

fastapi==0.111.0

### Pydantic Version

pydantic==2.7.1

### Python Version

3.10

**Comment by YuriiMotov** [comment]
This looks more like a question about using Pydantic V2 rather than FastAPI-related question.
I'll close it for now to keep the discussions clean.
If you’ve found a solution, please share it here.
If you still need help, feel free to let us know.
